### PR TITLE
Bug fix for new track id assignment + GPU memory leaks + tracker cleanup

### PR DESCRIPTION
**Pull request by shaikh58**
Previous logic increments an integer that represents the current track. However, this is brittle and has failed in at least 2 situations, and happens when a new track id is required by the tracker. This happens when the association score doesn't cross the min threshold for the given matching. This fix implements the latest track id as a set so that duplicates are not possible

Also deprecates the weight decay post processing param as this is rarely used and requires lots of extra computation i.e. a second pass through the visual encoder for the whole batch

EDIT (16 oct) : squeezing in GPU memory leak fix into this PR, and rebased off main since the keypoint dilate mask feature was merged into main after this PR was opened

EIT (17 October) : This also includes general tracker cleanup of deprecated features such as weight decay, Reid feature calculation, end_tracks in tracker_queue, and the persistent_tracking flag

### CODE REVIEW

**Comment by shaikh58** [comment]
Latest commit (on Oct 15) is just rebasing off current main (which also has the key point dilating mask feature)

File: dreem/inference/track_queue.py [hunk 1]
```diff
@@ -59,7 +59,7 @@ def __repr__(self) -> str:
             f"window_size={self.window_size}, "
             f"max_gap={self.max_gap}, "
             f"n_tracks={self.n_tracks}, "
-            f"curr_track={self.curr_track}, "
+            # f"curr_track={self.curr_track}, "
```
Comment (aaprasad) [inline dreem/inference/track_queue.py]: delete for clarity?

File: dreem/inference/tracker.py [hunk 2]
```diff
@@ -113,25 +114,25 @@ def track(
...
 
-        for frame in frames:
-            if frame.has_instances():
-                if not self.use_vis_feats:
-                    for instance in frame.instances:
-                        instance.features = torch.zeros(1, model.d_model)
-                    # frame["features"] = torch.randn(
-                    #     num_frame_instances, self.model.d_model
-                    # )
+        # for frame in frames:
```
Comment (aaprasad) [inline dreem/inference/tracker.py]: just delete this, can always go back to a previous commit if we need to get it back

File: dreem/inference/tracker.py [hunk 3]
```diff
@@ -168,9 +169,10 @@ def sliding_inference(
         # W: width.
 
         for batch_idx, frame_to_track in enumerate(frames):
-            tracked_frames = self.track_queue.collate_tracks(
-                device=frame_to_track.frame_id.device
-            )
+            self.device = frame_to_track.device
+            frame_to_track = frame_to_track.to("cpu")
```
[the remark this answers is not in the document]
Comment (aaprasad) [reply dreem/inference/tracker.py]: nvm i see below. Is there a better way to do this where we can just move it off the gpu after its been tracked, instead of moving it to the cpu and then moving it back to the gpu?

File: dreem/inference/tracker.py [hunk 4]
```diff
@@ -271,13 +276,18 @@ def _run_global_tracker(
...
 
-        reid_features = torch.cat([frame.get_features() for frame in frames], dim=0)[
-            None
-        ]  # (1, total_instances, D=512)
+        # reid_features = torch.cat([frame.get_features() for frame in frames], dim=0)[
+        #     None
+        # ]  # (1, total_instances, D=512)
 
         # (L=1, n_query, total_instances)
+        for instance in all_instances:
```
Comment (aaprasad) [inline dreem/inference/tracker.py]: see above. I feel like its expensive to move the instances to the gpu every frame

File: dreem/inference/tracker.py [hunk 5]
```diff
@@ -374,17 +384,17 @@ def _run_global_tracker(
         # reweighting hyper-parameters for association -> they use 0.9
 
         traj_score = post_processing.weight_decay_time(
-            asso_nonquery, self.decay_time, reid_features, window_size, query_ind
+            asso_nonquery, self.decay_time, None, window_size, query_ind
```
Comment (aaprasad) [inline dreem/inference/tracker.py]: why `None`? Can we just delete `reid_features` as an arg if its not being used anymore?

**Review by aaprasad — COMMENTED** [review]
looks mostly good to me. One main concern is it feels like we're moving to and from the gpu a lot more which could cause slowdowns. Should try and figure out a way to minimize gpu leakage w/o moving to/from gpu too much

File: dreem/inference/tracker.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shaikh58) [reply dreem/inference/tracker.py]: so currently once frame_to_track fwd pass is done and its moved to cpu, its out of scope. The new frame in the loop in sliding_inference is already on gpu since lightning moved it there right after the data loader. That is then moved to cpu to ensure none of the tq is on gpu. Cleanest option I see is to prevent lightning from moving the pred batch to gpu in the first place since we don't even do a fwd pass on the whole batch. Regarding speed, I didn't notice any slowdown on the handful of files I checked on, but didn't test thoroughly

File: dreem/inference/tracker.py (hunk 5 above)
Comment (shaikh58) [reply dreem/inference/tracker.py]: we can - this was a temp patch. Will remove it. Just need to pass the appropriate value into weight decay as it currently loop over features but doesn't actually use them

File: dreem/inference/tracker.py (hunk 2 above)
Comment (shaikh58) [reply dreem/inference/tracker.py]: removed

File: dreem/inference/track_queue.py (hunk 1 above)
Comment (shaikh58) [reply dreem/inference/track_queue.py]: deleted
